pkg/util/color/color.go

```diff
@@ -0,0 +1,165 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// This code originated in the github.com/golang/glog package.
+
+// Package color is a lightweight terminal color library. It is only compatible
+// with terminals which support ANSI escape codes.
+//
+// The API exposes two programming styles. In the first style, functions print
+// directly to the output stream:
+//
+//     color.Stdout(color.Red)
+//     fmt.Println("Red text")
+//     color.Stdout(color.Reset)
+//
+// In the second style, functions return a byte slice containing the escape
+// sequence:
+//
+//     fmt.Printf("%sRed text%s\n", color.StdoutProfile.Red(), color.StdoutProfile.Reset())
+//
+// Text destined for stderr should use color.Stderr and color.StderrProfile
+// instead.
+//
+// When the output stream is not a character device, or the TERM environment
+// variable indicates an unsupported terminal type, functions will print nothing
+// or return a nil byte slice.
+//
+// While we link the main CockroachDB binary against libncurses or libterminfo,
+// which bundle information about nearly every terminal ever produced, handling
+// colors for non-ANSI terminals is no small task. See, for example, the
+// difference between the setaf (set ANSI foreground) and setf (set foreground)
+// capabilities. Since non-ANSI terminals are few and far between, it's not
+// worth the trouble.
+package color
+
+import (
+	"os"
+	"runtime"
+	"strings"
+)
+
+// Code represents a terminal color code.
+type Code int
+
+// Color codes.
+const (
+	Red Code = iota
+	Yellow
+	Cyan
+	Light
+	Reset
+)
+
+// Profile defines escape sequences which provide color in terminals. Some
+// terminals support 8 colors, some 256, others none at all.
+type Profile [][]byte
+
+// For terminals with 8-color support.
+var profile8 = Profile{
+	// Keep these in the same order as the color codes above.
+	[]byte("\033[0;31;49m"),
+	[]byte("\033[0;33;49m"),
+	[]byte("\033[0;36;49m"),
+	[]byte("\033[2;37;49m"),
+	[]byte("\033[0m"),
+}
+
+// For terminals with 256-color support.
+var profile256 = Profile{
+	// Keep these in the same order as the color codes above.
+	[]byte("\033[38;5;160m"),
+	[]byte("\033[38;5;214m"),
+	[]byte("\033[38;5;33m"),
+	[]byte("\033[38;5;246m"),
+	[]byte("\033[0m"),
+}
+
+func (p Profile) get(code Code) []byte {
+	if p == nil {
+		return nil
+	}
+	return p[code]
+}
+
+// Red returns the escape sequence to produce red text.
+func (p Profile) Red() []byte { return p.get(Red) }
```

Comment:
I'm not convinced about this API. This looks like over engineering to me.
What's wrong with `profile[color]` like `p[Red]`? with a go map, so that you get the default value (empty byte slice) when the entry doesn't exist)

Reply:
TIL:

```
map[string]string(nil)["foo"]
```

does not panic. Will adjust accordingly.